Add data/compare and cuda vignettes (#255)
mrc-ide/dust: vignettes_src/cuda.Rmd

---
title: "dust: Running models on GPUs with CUDA"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{dust: Running models on GPUs with CUDA}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  error = FALSE,
  collapse = TRUE,
  comment = "#>"
)
lang_output <- function(x, lang) {
  cat(c(sprintf("```%s", lang), x, "```"), sep = "\n")
}
cc_output <- function(x) lang_output(x, "cc")
r_output <- function(x) lang_output(x, "r")
plain_output <- function(x) lang_output(x, "plain")
```

With the core approach in dust, you can run models in parallel efficiently up to the number of cores your workstation has available.  Getting more than 32 or 64 cores is hard though, and `dust` provides no multi-node parallelism (e.g., [MPI](https://en.wikipedia.org/wiki/Message_Passing_Interface)). Instead, we have developed a system for running `dust` models on GPUs (graphical processing units), specifically NVIDIA GPUs via CUDA (Compute Unified Device Architecture).

> This vignette is written in reverse-order, starting with how to run a model on a GPU, before covering how to write a `dust` model that can be run on a GPU. The reason for this is that we do not expect that people will write these directly; instead we expect that most people will use the [`odin.dust`](https://mrc-ide.github.io/odin.dust/) package to generate these interfaces automatically, without having to write as line of C++.

Reviewer: ```suggestion
This vignette is written in reverse-order, starting with how to run a model on a GPU, before covering how to write a `dust` model that can be run on a GPU. The reason for this is that we do not expect that people will write these directly; instead we expect that most people will use the [`odin.dust`](https://mrc-ide.github.io/odin.dust/) package to generate these interfaces automatically, without having to write out lines of C++.
```

## Principles

> * A GPU model can be run on the CPU
> *

Reviewer: ```suggestion
* A GPU model can be run on the CPU, and vice-versa.
* The same random number generator sequence will be used in both cases.
* Differences in results, if they exist, come from different precision/rounding between devices. When using `real_t = double` we get the same results.
* Just the model update can be defined for the GPU, and comparisons and shuffling will still happen on the CPU. Defining a comparison function is also possible, allowing a full particle filter run on a GPU.
```

> ## Running a model with GPU support

Reviewer: Add a section on run requirements? I think these would be:
- CUDA toolkit v10.2 or higher, v11.1 or higher preferred (compile time and run time)
- CUDA capable GPU (run time)
- nvidia drivers (run time). 

Check with `nvidia-smi`

The `sirs` model includes GPU support, and will be the focus of this vignette.  However, the installed version cannot be run directly on the GPU for a couple of reasons:

* this would complicate distribution of binaries as we'd depend on all systems having a copy of the CUDA runtime
* we would have to compile support for many possible GPU architectures at once
* it is very likely that we'd want to run the model in single precision (`float`) mode rather than double precision (`double`), and changing that requires re-compilation

So here, rather than using `dust::dust_example`, we compile the model and pass in arguments:

```{r}
path <- system.file("examples/sirs.cpp", package = "dust")
sirs <- dust::dust(path, gpu = TRUE, real_t = "float")
```

> Notice in compilation that `nvcc` is used to compile the model, rather than `g++`.  The additional option `-gencode=arch=compute_XX,code=sm_XX` was added by `dust` and will include the CUDA compute version supported by your graphics cards. You can use `dust::dust_cuda_options` to set additional options, passing in the return value for the `gpu` argument above.

Reviewer: ```suggestion
Notice in compilation that `nvcc` is used to compile the model, rather than `g++`.  The additional option `-gencode=arch=compute_XX,code=sm_XX` was added by `dust` and will include the CUDA compute version supported by the graphics cards found on the current system. You can use `dust::dust_cuda_options` to set additional options, passing in the return value for the `gpu` argument above.
```

Once compiled with GPU support, the static method `has_cuda` will report `TRUE`:

```{r}
sirs$public_methods$has_cuda()
```

and the static method `device_info` will report on the devices available:

```{r}
sirs$public_methods$device_info()
```

If you have more than one GPU, the `id` in the `devices` section will be useful for targeting the correct device.

The object is initialised as usual but with slight differences:

> * you will probably want a (much) larger number of particles to take advantage of your device

Reviewer: ```suggestion
* you will probably want a (much) larger number of particles to take advantage of your device. As a rule of thumb we would suggest at least 10000, but depending on model and card 10-100x this may improve performance (relative to the same on a multicore CPU). See below for more discussion of this.
```

* the `device_config` argument needs to be provided to indicate which device we are running on (or additional configuration)

```{r}
model <- sirs$new(list(), 0, 8192, device_config = 0L, seed = 1L)
```

Above, we also set the index to show just the daily incidence (number of new infections).

The model can be run on the CPU, as usual, and will be fairly slow here as this is a lot of particles:

```{r}
system.time(model$run(400))
```

To run the model on the device, pass `device = TRUE` to `run()`:

```{r}
model$reset(list(), 0L)
system.time(model$run(400, device = TRUE))
```

This is much faster! However, 8,000 particles is unlikely to saturate a modern GPU and (overhead-aside) this will run about as quickly for potentially a hundred thousand particles. For example running 2^16 (`r 2^16`) particles only takes a little longer

```{r}
model_large <- sirs$new(list(), 0, 2^17, device_config = 0L, seed = 1L)
system.time(model_large$run(400, device = TRUE))
```

Of course, _doing_ anything with all these particles is its own problem.

> The `run` and `simulate` methods both support a `device` function, which runs on the GPU if `TRUE`.  In addition, the `filter` method (for models with compare support) also supports a `device` argument, though this is typically used from the [`mcstate` interface](https://mrc-ide.github.io/mcstate/reference/particle_filter.html).

Reviewer: Add link to compare support vignette

  state_next[1] = I + n_SI - n_IR;
  state_next[2] = R + n_IR - n_RS;
  state_next[3] = (step % freq == 0) ? n_SI : state[3] + n_SI;
}
```

Note that this totally duplicates the logic from the CPU version (which was a method of the `sirs` object)

```cc
  void update(size_t step, const real_t * state,
              dust::rng_state_t<real_t>& rng_state,
              real_t * state_next) {
    real_t S = state[0];
    real_t I = state[1];
    real_t R = state[2];
    real_t N = S + I + R;

    real_t p_SI = 1 - exp(- shared->beta * I / N);
    real_t p_IR = 1 - exp(-(shared->gamma));
    real_t p_RS = 1 - exp(- shared->alpha);

    real_t n_SI = dust::distr::rbinom(rng_state, S, p_SI * shared->dt);
    real_t n_IR = dust::distr::rbinom(rng_state, I, p_IR * shared->dt);
    real_t n_RS = dust::distr::rbinom(rng_state, R, p_RS * shared->dt);

    state_next[0] = S - n_SI + n_RS;
    state_next[1] = I + n_SI - n_IR;
    state_next[2] = R + n_IR - n_RS;
    state_next[3] = (step % shared->freq == 0) ? n_SI : state[3] + n_SI;
  }
```

Differences include:

* the presence of the four auxiliary data elements (`internal_int`, `internal_real`, `shared_int` and `shared_real`)

> * the data types that vary across particles are a special `dust::interleaved<>` type, which prevents uncoalesced reads

Reviewer: ```suggestion
* the data types that vary across particles are a special `dust::interleaved<>` type, which prevents slow uncoalesced reads from global device memory
```

* All accesses into the `shared_int` and `shared_real` elements are now by position in the array, rather than the previous pointer/name based access

> * The `DEVICE` annotation

Reviewer: ```suggestion
* The `DEVICE` annotation, which compiles the function for use on the GPU
```

Debugging on a GPU is a pain, especially because there are typically many particles, and error recovery is not straightforward.  In addition, most continuous integration systems do not provide GPUs, so testing your GPU code becomes difficult.  To make this easier, `dust` allows running GPU code on the CPU - this will be typically slower than the CPU code, but allows easier debugging and verification that the model is behaving.  We use this extensively in `dust`'s tests and also in models built using `dust` that will run on the GPU.

To do this, compile the model with your preferred real type, but set the `gpu` argument to `FALSE`

```{r}
sirs_cpu <- dust::dust(path, gpu = FALSE, real_t = "float")
```

Note that above the model is compiled with `g++`, not `nvcc`. However, the "device" code is still compiled into the model.  We can then initialise this with and without a `device_config` argument

```{r}
model1 <- sirs_cpu$new(list(), 0, 10, seed = 1L)
model2 <- sirs_cpu$new(list(), 0, 10, device_config = 0L, seed = 1L)
```

And run the models using the "device" code and the normal CPU code:

```{r}
model1$run(10, device = FALSE)
model2$run(10, device = TRUE)
```

> These should be exactly the same, and this can form the basis of tests. Note that using `float` rather than `double` can throw up a few issues with algorithms, so it's worth checking with single-precision in your tests.

Reviewer: May want to note you can run with `R -d cuda-gdb` to get a debugger, and perhaps also link your blog post about gdb + valgrind?
